Align broadcast

crusaderky opened the pull request:
- Removed partial_align()
- Added exclude and indexes optional parameters to align() public API
- Added exclude optional parameter to broadcast() public API
- Added various unit tests to check that align() and broadcast() do not perform needless data copies
- broadcast() to automatically align inputs

Note: there is a failed unit test, TestDataset.test_broadcast_nocopy, which shows broadcast on dataset doing a data copy whereas it shouldn't. Could you look into it?

shoyer commented on xarray/core/alignment.py:
```diff
@@ -367,6 +391,7 @@ def _broadcast_dataset(ds):
 
 
 def broadcast_arrays(*args):
+    import warnings
```
oops! this is probably an indicator that nobody is relying on this deprecated functionality, though ;).

shoyer commented on xarray/core/alignment.py:
```diff
@@ -322,32 +319,59 @@ def broadcast(*args):
+        # Add excluded dims to a copy of dims_map
+        array_dims_map = dims_map.copy()
+        for dim in exclude:
+            try:
```
Maybe cleaner to check rather than doing the try/except: `if dim in array.dims`.

crusaderky replied:
That would cost double if the dim is present, and would be functionally identical... Up to you though

shoyer commented on xarray/core/alignment.py:
```diff
@@ -322,32 +319,59 @@ def broadcast(*args):
+                except ValueError:
+                    # dim not in var.dims
+                    try:
+                        del var_dims_map[dim]
```
I don't think this ever succeeds?

shoyer replied:
this is correct, I just didn't follow the logic correctly

shoyer replied:
Looking at how `dims_map` is initialized, we never add add variables in `exclude`  in. So now I'm not sure how `del` could fail.

crusaderky replied:
ah - you're right

shoyer commented on xarray/core/alignment.py:
```diff
@@ -322,32 +319,59 @@ def broadcast(*args):
-            data_vars[k] = ds.variables[k].expand_dims(dims_map)
+            var = ds.variables[k]
+            # Add excluded dims to a copy of dims_map    
+            var_dims_map = dims_map.copy()
```
maybe better to write a helper function rather than writing the twice

crusaderky replied:
You're right I'll fix it

shoyer commented on xarray/test/test_dataarray.py:
```diff
@@ -1636,9 +1690,45 @@ def test_broadcast_arrays(self):
+        x2, y2 = broadcast(x, y)
+        self.assertDataArrayIdentical(expected_x2, x2)
+        self.assertDataArrayIdentical(expected_y2, y2)
+        assert x2.data is x.data
```
It's a little tricky to verify that two arrays share the same data. See here for discussion:
http://stackoverflow.com/questions/11286864/is-there-a-way-to-check-if-numpy-arrays-share-the-same-data

In many cases, looking at `base` works. I wrote a little utility to help with that:
https://github.com/pydata/xarray/blob/765f17d3b31af1228c6c8def8e108f62a27d7260/xarray/test/__init__.py#L222

So maybe this would work: `source_ndarray(x2.data) is source_ndarray(x.data)`

`numpy.shares_memory` is a clean way to do this, but is only available in numpy 1.11+.

shoyer commented on xarray/test/test_dataarray.py:
```diff
@@ -1636,9 +1690,45 @@ def test_broadcast_arrays(self):
+        self.assertDataArrayIdentical(expected_y2, y2)
+
+    def test_broadcast_arrays_nocopy(self):
+        # Test that input data is not copied over in case no alteration is needed
```
The guarantee is actually a little stronger than this -- the output arrays should still be views of the original data (no copies), even if broadcasting *was* required.

Alignment does require a copy if the indexes differ.

crusaderky replied:
I changed all tests to use source_ndarray(). I'm still getting a failure in a dataset alignment that seems to be performing an unwarranted copy though. Could you take a look?

shoyer commented on xarray/test/test_dataset.py:
```diff
@@ -1056,6 +1056,35 @@ def test_align(self):
+    def test_align_nocopy(self):
+        x = Dataset({'foo': DataArray([1, 2, 3], coords={'x': [1, 2, 3]})})
+        y = Dataset({'foo': DataArray([1, 2], coords={'x': [1, 2]})})
+        x2, y2 = align(x, y, join='outer')
```
You need to add `copy=False` here -- otherwise we do a copy by default.

shoyer commented:
can you add a brief note to "what's new"?

I'm going to issue another bug fix release shortly, would be nice to get this in :)

crusaderky commented:
Updated what's new, fixed unit test bug, and added auto-align to concat().
I plan to add auto-broadcast to it as well eventually, but I realised it's not trivial as when doing it on datasets you must perform it pairwise. I'll leave it for later.

shoyer commented on doc/whats-new.rst:
```diff
@@ -24,6 +24,11 @@ Enhancements
 - New documentation on :ref:`panel transition`. By
   `Maximilian Roos <https://github.com/MaximilianR>`_.
 
+- Added exclude and indexes optional parameters to align(), and exclude optional parameter
```
Use back ticks to denote code:

   ``exclude``

and sphinx directives to like to function docstrings:

    :py:func:`~xarray.align`

shoyer commented on doc/whats-new.rst:
```diff
@@ -24,6 +24,11 @@ Enhancements
 
+- Added exclude and indexes optional parameters to align(), and exclude optional parameter
+  to broadcast(). By `Guido Imperiale <https://github.com/crusaderky>`_.
+- broadcast() and concat() will now auto-align inputs.
```
add "using `join=outer`'

shoyer commented:
OK, this looks great to me! I especially appreciate all the extra test coverage.

I will merge once the CI tests are green.

shoyer commented:
Actually, tests were passing on the previous commit (pre doc fixes) so I decided to go ahead.

Thanks Guido!
